lib/plenario/actions/meta_actions.ex

```diff
@@ -1,424 +1,388 @@
   end
 
-  def disapprove(_, admin, _), do: {:error, "#{admin.name} is not an admin"}
+  def guess_field_types(meta) do
+    %HTTPoison.Response{body: body} = HTTPoison.get!(meta.source_url)
+    path = "/tmp/#{meta.slug}.#{meta.source_type}"
+    File.write!(path, body)
+
+    cnt = 1_001
+    first_thousand =
+      case meta.source_type do
+        "csv" -> parse_csv(path, cnt)
+        "tsv" -> parse_tsv(path, cnt)
+        "json" -> parse_json(path, cnt)
+        "shp" -> parse_shapefile(path, cnt)
+      end
 
-  @doc """
-  Handle the state transition of the Meta from whatever to erred
-  """
-  @spec mark_erred(meta :: %Meta{}, admin :: %User{}, message :: String.t()) ::
-          {:ok, %Meta{} | :error, Ecto.Changeset.t()}
-  def mark_erred(meta, %User{is_admin: true} = admin, message) do
-    msg = "An error occurred regarding your data set:\n\n" <> message
-
-    AdminUserNoteActions.create_for_meta(
-      msg,
-      admin,
-      meta.user,
-      meta,
-      true
-    )
+    guesses =
+      for row <- first_thousand do
+        for {key, value} <- row do
+          cond do
+            check_boolean(value) -> {key, "boolean"}
+            check_integer(value) -> {key, "integer"}
+            check_float(value) -> {key, "float"}
+            check_date(value) -> {key, "timestamptz"}
+            check_json(value) -> {key, "jsonb"}
+            true -> {key, "text"}
+          end
+        end
+      end
+      |> List.flatten()
+
+    counts =
+      Enum.reduce(guesses, %{}, fn x, acc -> Map.update(acc, x, 1, &(&1 + 1)) end)
+      |> Enum.into([])
+
+    maxes =
+      Enum.reduce(counts, %{}, fn {{k, _}, c}, acc ->
+        curr = Map.get(acc, k, 0)
+        if c > curr do
+          Map.update(acc, k, c, &(&1 - &1 + c))
+        else
+          acc
+        end
+      end)
+
+    col_types =
+      Enum.reduce(maxes, %{}, fn {col, max_c}, acc ->
+        {{_, type}, _} = Enum.find(counts, fn {{k, _}, c} -> "#{k}" == "#{col}" and c == max_c end)
+        Map.merge(acc, %{col => type})
+      end)
+
+    col_types
+  end
 
-    Meta.mark_erred(meta)
-    |> Repo.update()
+  defp parse_csv(filename, count) do
+    File.stream!(filename)
+    |> Enum.take(count)
+    |> CSV.decode!(headers: true)
   end
 
-  def mark_erred(_, admin, _), do: {:error, "#{admin.name} is not an admin"}
+  defp parse_tsv(filename, count) do
+    File.stream!(filename)
+    |> Enum.take(count)
+    |> CSV.decode!(headers: true, separator: ?\t)
+  end
 
-  @doc """
-  Handle the state transition of the Meta from erred to ready
-  """
-  @spec mark_fixed(meta :: %Meta{}, admin :: %User{}, message :: String.t()) ::
-          {:ok, %Meta{} | :error, Ecto.Changeset.t()}
-  def mark_fixed(meta, %User{is_admin: true} = admin, message) do
-    msg = "Data set error fixed:\n\n" <> message
-
-    AdminUserNoteActions.create_for_meta(
-      msg,
-      admin,
-      meta.user,
-      meta,
-      true
-    )
+  defp parse_json(filename, count) do
```

Comment:
These parsing functions are really similar to those used in the worker. I'll make a note to abstract these out to their own modules when I get the chance.

Reply:
Needs an issue.